Use native URL in several media modules.

sgomes opened the pull request:
This PR replaces usage of the deprecated browserified node `url` module (and `qs`) with the native `URL` and `URLSearchParams` across several media-related modules. It's one of many PRs working towards dropping 3rd-party libraries for URL and URL parameter handling and replacing them with native functionality.

It also reduces usage of `lodash`.

Note: If I added you and you feel you're not the right person to review these changes, I apologise for that; please feel free to ignore or remove yourself. I just went with the default GitHub suggestions since I had no idea who to add 🤷‍♂

#### Changes proposed in this Pull Request

* Replace usage of `url` with the native `URL`/`URLSearchParams` and `lib/url`.
* A number of changes to make the linter happy with the files I touched.

#### Testing instructions

The unit tests seem to cover the changes pretty thoroughly.

dmsnell commented on client/components/gravatar/index.jsx:
```diff
@@ -46,22 +37,29 @@ export class Gravatar extends Component {
+		} catch {
+			// Invalid URL; return unmodified.
+			return imageURL;
+		}
```
Particularly since we'll need to do this everywhere we use `URL` can we create a wrapper function that handles the exception? I'm not sure if there's anywhere we would need to differentiate an exception from a falsey value, and I'm predicting that people will call `new URL()` without the try/catch.

```js
const parseURL = s => {
	try {
		return new URL( s );
	} catch ( e ) {
		return null;
	}
}
```

```js
const parsedURL = parseURL( imageURL );
if ( ! parsedURL ) {
	return imageURL;
}
```

sgomes replied:
I'm afraid it's not as simple as that. The URL parsing takes a second, optional parameter, for a base URL. This is because the `URL` object only supports absolute URLs, and so it needs a base URL to complete any partial URL that may be provided.

As such, we may want different behaviours, depending on whether or not that 2nd parameter is provided.

In this case, it's not, which means we assume that the provided URL will be an absolute URL and we want it to fail if not.

sgomes replied:
In fact, let me explain things a bit deeper, since `URL` isn't trivial.

`URL` only handles absolute URLs. It's not able to represent a partial URL at all, even if just the protocol is missing (making it a scheme-relative URL). There is a second parameter in the `URL` constructor, which allows you to provide a base URL, which is used to resolve the missing bits in the first parameter. That is, if you have a path-absolute URL, for example, and provide a base URL, it'll take the protocol and domain from the base URL, while keeping the path from the first parameter:

```js
const url = new URL( '/my/path', 'https://wordpress.com/other/path');
url.href === 'https://wordpress.com/my/path';
```

However, the deprecated `parse` and `format` from `node`'s `url` are one of those magic APIs that handle all types of URLs without complaining (or without erroring out when they should, in many cases), which means that we have a lot of code that expects to be able to parse and format partial URLs of various sorts.

As such, I added two functions to `lib/url`: `determineUrlType` and `format`. `determineUrlType` determines the URL type for a provided URL: absolute (`http://wordpress.com/path`), scheme-relative (`//wordpress.com/path`), path-absolute (`/path`), path-relative (`path` or `../path`), or invalid (any invalid URL). `format` normalises a URL, optionally downcasts it to a lesser type (e.g. from absolute to scheme-relative) and outputs it as a string.

With all this in mind, I could definitely take the approach of reimplementing something similar to `node` `url`'s `parse` and `format` based on `URL`. It would be the easiest way to migrate code, with as few changes as possible. However, it seems to me that it could perpetuate some of mistakes of the `node` API:
- It's hard to tell what types of URL a piece of code that makes use of the API is expecting.
- It's hard to limit the types of URL you want to handle.
- It's easy to make mistakes, in thinking you're only handling absolute URLs, when you're actually handling everything.

I'm happy to discuss different ways of handling things more easily, without as much trying and catching going on. The other methods in `lib/url` (such as `addQueryArgs`, `omitURLParams`, etc.) already provide some help, by offering easier, focused high-level APIs, but I haven't come up with a good API for the general use-case that doesn't suffer from the same issues in `node`'s `url`.

dmsnell commented on client/lib/media/utils.js:
```diff
@@ -111,7 +112,7 @@ export function getFileExtension( media ) {
 		let filePath;
 		if ( isUri( media ) ) {
-			filePath = urlLib.parse( media ).pathname;
+			filePath = new URL( media, BASE_PARSE_URL ).pathname;
```
here's that risk of throwing an uncaught exception

in fact, I wonder here if this change warrants more refactoring, because it looks like `isUri` would be obviated by a robust URL parser

```js
if ( isString ) {
	const filePathUrl = parseURL( media, BASE_PARSE_URL );
	const filePath = filePathUrl ? filePathUrl.pathname : media;
	
	extension = path.extname( filepath ).slice( 1 );
}
```

sgomes replied:
There is no risk here, because we already used `isUri` to determine if the string is a valid URL. Since we're using a base path, this won't throw on partial URLs, and this won't fail at all.

`isUri` provides different checks which are needed in this case. In particular, it looks for invalid characters in the URL string which would be otherwise ignored. Since this method is trying to handle both URLs and filesystem locations, we need the stricter validation up top, and can't just rely on `URL`'s validations.

dmsnell commented on client/lib/media/utils.js:
```diff
@@ -122,8 +123,13 @@ export function getFileExtension( media ) {
+			extension = path.extname( pathname ).slice( 1 );
+		} catch {
+			return '';
+		}
```
same here…

```js
} else {
	const parsedPath = parseURL( media.URL || media.file || media.guid || '', BASE_PARSE_URL );
	const pathname = parsedPath.?pathname || '';

	extension = path.extname( pathname ).slice( 1 );
}
```

not sure exactly what behavior we're trying to handle in the `catch` statement

sgomes replied:
This else block handles the situation where we weren't passed a string, but rather an object with one of several well-known properties. It tries to compose a URL from that and look for the extension, and returns an empty string otherwise.

It will return an empty string if it can't find any of the properties, or if there's no extension. It will throw, be caught, and return an empty string if the URL is invalid.

dmsnell commented on client/lib/media/utils.js:
```diff
@@ -186,7 +192,7 @@ export function getMimeType( media ) {
 
 	extension = extension.toLowerCase();
-	if ( MimeTypes.hasOwnProperty( extension ) ) {
+	if ( Object.keys( MimeTypes ).includes( extension ) ) {
```
what's this change for? also, it looks like we're changing what might be a more efficient method with one that creates a new array of object keys then runs a slower method on that new array…?

sgomes replied:
The linter has now started complaining about these checks; this seemed like the most correct way of performing the same check. I could instead do `extension in MimeTypes`, but that could potentially fail if `extension` happens to be a string that matches one of the object built-ins (because `in` traverses the prototype chain).

dmsnell replied:
> As for performance, it would seem like a questionable micro-optimisation to opt for a different syntax purely for performance reasons here.

It caught my attention because it was an active change more than because it's slower. As in, sometimes a change by itself doesn't flag anything, but a change in the slower direction, especially from you might rouse suspicion 😉 

> The linter has now started complaining about these checks;

For those who are less informed, and less enthusiastic, what does the linter have to complain about this?

This is quite idiomatic JavaScript and I instantly knew what it was checking. In contrast, I had to open the JS docs to see if `Object.keys()` behaves like `getOwnPropertyNames()` or `for…in` or `for…of`.

What are your thoughts here? If the linter hadn't complained would you have made this change?

sgomes replied:
Changed back, as the linter no longer seems to care.

dmsnell commented on client/lib/resize-image-url/index.js:
```diff
@@ -61,32 +61,35 @@ const scaleByFactor = value => value * IMAGE_SCALE_FACTOR;
+
+	const parsedUrl = new URL( imageUrl, 'https://__domain__.invalid' );
+
+	if ( ! REGEXP_VALID_PROTOCOL.test( parsedUrl.protocol ) ) {
```
exception risk, and unnecessary RegExp after the change…

```js
const url = parseURL( imageUrl, 'https://__domain__.invalid' );

if ( ! ( url && ( 'https:' === url.protocol || 'http:' === url.protocol ) ) ) {
	return imageUrl;
}
```

sgomes replied:
There's no exception risk here, because we already tested above that we don't have an invalid URL (we'd get `URL_TYPE.INVALID` on the `urlType` if that were the case).

dmsnell commented on client/lib/resize-image-url/index.js:
```diff
@@ -61,32 +61,35 @@ const scaleByFactor = value => value * IMAGE_SCALE_FACTOR;
 	}
 
-	parsedUrl.query = omit( parsedUrl.query, SIZE_PARAMS );
+	SIZE_PARAMS.forEach( param => parsedUrl.searchParams.delete( param ) );
```
noting that in addition to removing `lodash` here we are also introducing mutation into the variable which is referenced through multiple lines of this function. this may be fine, but it's something we like to avoid in general so that we don't have to interpret the code as much in our heads when reading, don't have to pay attention to what might have changed it.

faster than the original, slower than your change…

```js
const withoutSizes = SIZE_PARAMS.reduce( ( finalUrl, nextSize ) => {
	finalUrl.searchParams.delete( nextSize );
	return finalUrl;
}, new URL( url.toString() );
```

This is only marginally different but by cloning the URL first we can avoid that mutation. Could also keep the `forEach` and clone separately if on purity grounds we don't like mutating inside a `reduce`

```js
const withoutSizes = new URL( url.toSring() );
SIZE_PARAMS.forEach( param => withoutSizes.searchParams.delete( param ) );
```

This way although the mutation is there it's isolated and colocated with the initial assignment.

sgomes replied:
I don't really see the issue with mutating an object in which you're preparing a result. Is it the fact that it's named `parsedUrl`? I'll change the name to `resultUrl` to make the intention clearer.

dmsnell commented on client/lib/resize-image-url/index.js:
```diff
@@ -104,23 +107,22 @@ export default function resizeImageUrl( imageUrl, resize, height, makeSafe = tru
+
+	for ( const key of Object.keys( mapped ) ) {
+		parsedUrl.searchParams.set( key, mapped[ key ] );
+	}
```
same idea as above with mutation. personally I end up with more confidence and less anxiety if I can see that a value is set once and I don't have to hunt the rest of the function to see if it's changed later, or could be changed in different ways depending on runtime characteristics.

```js
const resizedUrl = new URL( withoutSizes.href );
Object.keys( mapped ).forEach( key => resizedUrl.searchParams.set( key, mapped[ key ] ) );
```

at which point we may think about creating a new name altogether for what we're doing. it takes me a second glance to understand what we're trying to do here

```js
const withSearchParams = ( baseUrl, params ) => {
	const url = new URL( baseUrl.href );
	Object.keys( params ).forEach( name => url.searchParams.set( name, params[ name ] ) );
	return url;
}

const withoutSearchParams = ( baseUrl, paramNames ) => {
	const url = new URL( baseUrl.href );
	paramNames.forEach( name => url.searchParams.delete( name ) );
	return url;
}
```

```js
const withoutSizes = withoutSearchParams( parsedUrl, SIZE_PARAMS );
```

```js
return withSearchParams( withoutSizes, mapped ).href;
```

`mapped` may not be the most ideal name either, but I'm not trying to push too far into this specific module, just suggesting that having those names might more quickly indicate the intention of the loops and mutation, as well as give us a more concrete unit to test.

sgomes replied:
Again, I think it's pretty idiomatic to create a result value and mutate it as the function progresses. Hopefully renaming the variable made it clearer, but let me know if you'd still like me to use different `const`s for each individual stage of processing.

dmsnell replied:
this shouldn't block the PR

dmsnell reviewed (commented):
Thanks for the poke; this is a great change. In the past I had some feeling like we avoided this for compatibility reasons, but I can't find what that would be.

Getting rid of some of that `lodash` and adding the `?.` is a nice refreshment.

I'm pretty sure there's strong value in creating a wrapper to handle that exception and I'd like your thoughts on that. There's already code in here that risks throwing the uncaught exceptions I was immediately worried about when I saw the first change, and we should probably expect people to follow suit.

sgomes commented:
> Thanks for the poke; this is a great change. In the past I had some feeling like we avoided this for compatibility reasons, but I can't find what that would be.
> 
> Getting rid of some of that `lodash` and adding the `?.` is a nice refreshment.
> 
> I'm pretty sure there's strong value in creating a wrapper to handle that exception and I'd like your thoughts on that. There's already code in here that risks throwing the uncaught exceptions I was immediately worried about when I saw the first change, and we should probably expect people to follow suit.

Thank you for your excellent review, @dmsnell! I think it uncovered some fundamental questions as to how we want to be doing URL handling going forward. `node`'s deprecated APIs have too much magic, and `URL` is too limited, which means we need to build on top of `URL` to provide the missing functionality. I tried to do that with a couple of functions, but it may still be too low-level and work-intensive (details [here](https://github.com/Automattic/wp-calypso/pull/37643#discussion_r347350110))

I'll give a better API design some thought over the next few days, but if you come up with something please let me know!

dmsnell commented:
@sgomes thanks for you thoughtful responses to my comments. It's clear you had considered them all before I wrote them down.

> I'll give a better API design some thought over the next few days

One of the things that stands out to me especially after reading some of the explanations about `node`'s API falls back to the value that we're bringing here. If we get rid of some of our wrapper functions but then still rely on `isUrl` and other behaviors that could/should be wrapped up in those helpers, have we made it any harder to mess up in the code and have we removed any sizable enough dependencies?

Since you're working on a deprecated module it makes sense if we have to get that out of there.

Maybe what I'm trying to say is that I sympathize with the conundrum of dealing with URLs. It's something I've dealt with before and hope that someday through our use of [type aliases](https://github.com/Automattic/wp-calypso/blob/master/client/types.ts#L7) or even of stronger types we can make those distinctions clearer.

sgomes commented:
> One of the things that stands out to me especially after reading some of the explanations about `node`'s API falls back to the value that we're bringing here. If we get rid of some of our wrapper functions but then still rely on `isUrl` and other behaviors that could/should be wrapped up in those helpers, have we made it any harder to mess up in the code and have we removed any sizable enough dependencies?

I'm generally removing all of them as I go, with `isUrl` as the only exception so far. It seemed likeit didn't require immediate attention since it's just ~500 bytes (compressed), it's entirely based on RFCs, and its functionality isn't easily reproduced. `qs`, on the other hand, is 3.5KB, and the deprecated `url` module is 9KB, and they're both much easier to replace with `URL` + `lib/url`.

> Maybe what I'm trying to say is that I sympathize with the conundrum of dealing with URLs. It's something I've dealt with before and hope that someday through our use of [type aliases](https://github.com/Automattic/wp-calypso/blob/master/client/types.ts#L7) or even of stronger types we can make those distinctions clearer.

The new functions I added in `lib/url` are fully typed, which does help a little bit with e.g. `format`. If I were to come up with a different API that were to take URL description objects like `url`'s `format`, there would be more that could be done there too.

It won't possible to type the whole thing, though, since URLs are often strings, and you can't use the type system to infer e.g. if a string represents an absolute URL or a relative one.

dmsnell commented:
> you can't use the type system to infer e.g. if a string represents an absolute URL or a relative one.

true, but should we find the need we can use it to enforce those things. there are ways to create types that, for example, can only be produced inside a given function, or types that carry their name around but not necessarily the runtime overhead ([for example](https://github.com/gcanti/newtype-ts))

sgomes commented:
I'm placing this PR on hold while looking into `@wordpress/url`, as work there may influence this PR.

sgomes commented:
I'm placing this PR back in review, as we came to the conclusion that Gutenberg and Calypso have different requirements (with Gutenberg needing support for PHP-like query arg serialisation), and so it's best to continue with a separate `lib/url` in Calypso.

aduth commented:
> (with Gutenberg needing support for PHP-like query arg serialisation), and so it's best to continue with a separate `lib/url` in Calypso.

Does Calypso not have need for this? Isn't it calling to the core REST API in some cases?

sgomes commented:
> Does Calypso not have need for this? Isn't it calling to the core REST API in some cases?

`qs` is definitely being used in some places, so it's possible. Unfortunately, it's hard to tell without a full audit of the codebase, because these advanced features are enabled by default in `qs`, and it's often hard to tell trivial usage apart from advanced usage. I haven't personally come across any such usage yet, but I haven't dived too deeply into API usage.

In any case, I think this PR stands by itself, as it's mostly dealing with moving away from `url`, rather than `qs`. And if at some point we do discover that we need PHP-like serialisation in Calypso, I'd be more than happy to handle the migration to `@wordpress/url` myself, perhaps even adding to it with code from `lib/url` 🙂

aduth commented:
> In any case, I think this PR stands by itself, as it's mostly dealing with moving away from `url`, rather than `qs`. And if at some point we do discover that we need PHP-like serialisation in Calypso, I'd be more than happy to handle the migration to `@wordpress/url` myself, perhaps even adding to it with code from `lib/url` 🙂

Thanks for clarifying! The previous comment stood out to me in that I would hope for there not to be differing fundamental goals. I would love any opportunity to consolidate these approaches. Based on what you're saying here, I sense it is safe to continue this discussion separate from this pull request.

sgomes commented:
Rebased on master to pick up node 12 changes.

sgomes commented:
Rebased on master.

sgomes committed:
Use native URL in several media modules.

sgomes commented:
(Did not mean to close this PR; did a bad force push)

sgomes commented:
@dmsnell , @Automattic/team-calypso : I reimplemented this PR using the new higher-level methods I added to `lib/url` a while back. This is much closer to the original code now, and should be easier to review.

Any chance of an approval? Thank you!

sgomes committed:
Revert to use `hasOwnProperty`.

sgomes committed:
Rename parsedUrl in resizeImageUrl.

dmsnell commented on client/components/gravatar/index.jsx:
```diff
@@ -46,22 +38,23 @@ export class Gravatar extends Component {
 		imageURL = imageURL || 'https://www.gravatar.com/avatar/0';
-		const parsedURL = url.parse( imageURL );
-		const query = parse( parsedURL.query );
+		const parsedURL = getUrlParts( imageURL );
```
fun fact, in these cases we can also perform our "deletion" in a destructuring assignment

```js
const { search, ...parsedURL } = getUrlParts( imageURL );
```

Just provides one way to remove one more thing to keep track of (what properties on `parsedURL` remain and which have been or may have been deleted).

sgomes replied:
Good idea, thanks! I thought the linter would complain about `search` not being used anywhere, but that doesn't appear to be the case 👍

dmsnell commented on client/components/gravatar/index.jsx:
```diff
@@ -46,22 +38,23 @@ export class Gravatar extends Component {
 
-		parsedURL.search = stringify( query );
-		return url.format( parsedURL );
+		return getUrlFromParts( parsedURL ).href;
```
well this reads much better to me than the previous code - I trust it more knowing it's being properly parsed and returned.

dmsnell commented on client/lib/media/test/actions.js:
```diff
@@ -75,14 +75,15 @@ describe( 'MediaActions', () => {
 		window.FileList.prototype[ Symbol.iterator ] = Array.prototype[ Symbol.iterator ];
-		window.URL = { createObjectURL: sandbox.stub() };
+		savedCreateObjectURL = window.URL.createObjectURL;
+		window.URL.createObjectURL = sandbox.stub();
```
So this was here before, but why is it here and why do we have to stub it? I don't see anything obvious in the `media` tests explaining why. Is this part of the new URL parsing? part of the old? what happens if we don't stub it out? is it missing from `jsdom`?

sgomes replied:
The `createTransientMedia` function in `client/lib/media/utils.js` uses `URL.createObjectURL`, which isn't available in Node's `URL`. This is likely because it's part of the [File API](https://w3c.github.io/FileAPI/) and not the [URL standard](https://url.spec.whatwg.org/), even though it gets added as a static method to `URL`.

The reason why the stub was changed in this PR is because it was an incomplete implementation of `URL`. It became a simple object with a `createObjectUrl` method, and nothing else. This way, the full `URL` implementation is preserved (which is needed, since the new code makes full use of `URL`), and we still keep the stub for the `createObjectURL` method.

dmsnell commented on client/lib/media/utils.js:
```diff
@@ -488,6 +488,8 @@ export function canUserDeleteItem( item, user, site ) {
  * @param {number} quality extracted image quality
  */
 export function canvasToBlob( canvas, callback, type, quality ) {
+	const { HTMLCanvasElement, Blob, atob } = window;
```
this change definitely stands out. was this necessary or a separate fix? if we were to have to rollback the URL changes, would we also want this to rollback?

sgomes replied:
This is due to linter config changes. It used to be ok to refer to these directly, but that's no longer allowed. I decided to destructure window here, instead of prepending `window.` to every usage of these in the method.

dmsnell commented on client/lib/resize-image-url/index.js:
```diff
@@ -25,7 +25,7 @@ const REGEXP_VALID_PROTOCOL = /^https?:$/;
  */
 const IMAGE_SCALE_FACTOR =
-	get( typeof window !== 'undefined' && window, 'devicePixelRatio', 1 ) > 1 ? 2 : 1;
+	( typeof window !== 'undefined' && window?.devicePixelRatio ) > 1 ? 2 : 1;
```
before the change we always returned `1` but now we might return `false` if `window` is undefined. the result will be the same, but `false > 1` makes me pause to make sure

what about leaving the `> 1` inside the parens?

```js
( typeof window !== 'undefined' && window?.devicePixelRatio > 1) ? 2 : 1;
```

that way the whole subexpression is either `false` or `true` and that feeds into the ternary as a single type instead of as one of two

sgomes replied:
> before the change we always returned `1` but now we might return `false` if `window` is undefined.

I just want to be clear that will definitely not be the case, because of what you say afterwards:

> the result will be the same, but false > 1 makes me pause to make sure

I do agree that's not a great practice, though. Changed to your suggestion (minus the parens, which get removed by `prettier`) 👍

dmsnell replied:
> before the change we always returned 1 

poor wording on my part - I meant in the conditional part of the expression.

dmsnell commented on client/lib/resize-image-url/index.js:
```diff
@@ -73,20 +73,21 @@ export default function resizeImageUrl( imageUrl, resize, height, makeSafe = tru
-	const parsedUrl = parse( imageUrl, true, true );
-	if ( ! REGEXP_VALID_PROTOCOL.test( parsedUrl.protocol ) ) {
+	const resultUrl = getUrlParts( imageUrl );
+	delete resultUrl.search;
```
same fun fact as above; I'm sure you know this already just pointing it out

```js
const { search, ...resultUrl } = getUrlParts( imageUrl );
```

sgomes replied:
Done, thanks!

dmsnell reviewed (approved):
@sgomes although I left a few comments I have no problem with you merging this as-is. I figure that if you want to respond to them you can do so and merge since none of them are big changes and you can test and decide well enough 😄 

this is a good improvement. thanks for taking the care to implement it, adjust the framework, and interact in the reviews

sgomes committed:
Some readability fixes after review.

sgomes commented:
> @sgomes although I left a few comments I have no problem with you merging this as-is. I figure that if you want to respond to them you can do so and merge since none of them are big changes and you can test and decide well enough 😄
> 
> this is a good improvement. thanks for taking the care to implement it, adjust the framework, and interact in the reviews

Woohoo! I addressed all your comments through either fixes or comments of my own, and I think this is in a good state to ship. Will wait for the tests and then merge.

Thanks for the review and all your great suggestions, @dmsnell, and for pushing me to look for a better API for `lib/url`! 🙇

sgomes commented:
The E2E test failures seem unrelated. Merging.
